### PR TITLE
V2

### PR DESCRIPTION
**Pull request by AndreyLevchenko**

### CODE REVIEW

File: router/router.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+	stopTicker  chan struct{}
+	cfgfile     string
+	aquaServer  string
+	outputs     map[string]outputs.Output
+	inputRoutes map[string]*routes.InputRoute
+	templates   map[string]data.Inpteval
+}
+
+var (
+	errNoOutputs  = errors.New("there aren't started outputs")
```
Comment (afdesk) [inline router/router.go:48]: unused variable

File: router/router.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+	initCtx       sync.Once
+	routerCtx     *Router
+	baseForTicker = time.Hour
+
+	osStat = os.Stat
```
Comment (afdesk) [inline router/router.go:53]: unused variable

File: router/router.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+	ctx.cfgfile = cfgfile
+	ctx.outputs = map[string]outputs.Output{}
+	ctx.inputRoutes = map[string]*routes.InputRoute{}
+	ctx.templates = map[string]data.Inpteval{}
+	ctx.ticker = nil
+
+	err := ctx.load()
+	if err != nil {
+		return err
```
Comment (afdesk) [inline router/router.go:91]: this error will be lost.

File: router/router.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+		if err != nil {
+			return err
+		}
+
+		if resp.StatusCode > 399 {
+			return errors.New(fmt.Sprintf("can not connect to %s, response status is %d", template.Url, resp.StatusCode))
+		}
+
+		b, err := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
```
Comment (afdesk) [inline router/router.go:161]: I'd transfer this `defer` call after the check of error.

File: router/router.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+	}
+}
+
+func (ctx *Router) handle(in []byte) {
+	for routeName := range ctx.inputRoutes {
+		ctx.HandleRoute(routeName, in)
+	}
+}
+func BuildAndInitOtpt(settings *OutputSettings, aquaServerUrl string) outputs.Output {
+	var plg outputs.Output
```
Comment (afdesk) [inline router/router.go:291]: I'd transfer to line 305 (before `switch` and using).

File: routes/routes.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+type Plugins struct {
+	AggregateIssuesNumber   int    `json:"aggregate-issues-number"`
+	AggregateIssuesTimeout  string `json:"aggregate-issues-timeout"`
+	PolicyShowAll           bool   `json:"policy-show-all"`
+	AggregateTimeoutSeconds int
+}
+
+func (route *InputRoute) IsSchedulerRun() bool {
+	return route.Scheduling != nil
```
Comment (afdesk) [inline routes/routes.go:20]: we stop `Scheduling` via close a channel. Will `route.Scheduling` be `nil` after 'StopScheduler'?

File: ui/backend/dbservice/getplgnstats.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package dbservice
+
+import (
+	"strconv"
+
+	hookDbService "github.com/aquasecurity/postee/dbservice"
+	bolt "go.etcd.io/bbolt"
+)
+
+func GetPlgnStats() (r map[string]int, err error) {
```
Comment (afdesk) [inline ui/backend/dbservice/getplgnstats.go:10]: There are not locks for getting stat info, is it right?

**Review by afdesk — CHANGES_REQUESTED** [review]

File: Dockerfile.ui [hunk 8]
```diff
@@ -0,0 +1,34 @@
+FROM node:latest as vuebuilder
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply Dockerfile.ui:1]: done

File: Dockerfile.ui [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+
+FROM golang:1.16-alpine as gobuilder
+
+ADD . /webhook
+WORKDIR /webhook/ui/backend
+RUN go build -o posteeui
+
+FROM alpine
+EXPOSE 8000
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply Dockerfile.ui:16]: yes there is `POSTEE_UI_PORT` environment variable

File: data/types_test.go [hunk 10]
```diff
@@ -25,3 +25,28 @@ func TestGetUniqueId(t *testing.T) {
...
+	for _, test := range tests {
+		si := &ScanImageInfo{
+			Image:    test.image,
+			Registry: test.registry,
+			Digest:   test.digest,
+		}
+		if got := si.HasId(); got != test.expectedResult {
+			t.Errorf(`for scan (image: "%s",registry: "%s",digest: "%s") HasUniqueId()== %t, want %t`, test.image, test.registry, test.digest, got, test.expectedResult)
+		}
+
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply data/types_test.go:50]: done

File: msgservice/msghandling.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		prnInputLogs("Input %s... doesn't match a REGO rule: %s", input, route.Input)
+		return
+	}
+
+	if err := scan.init(input); err != nil {
+		log.Println("ScanService.Init Error: Can't init service with data:", input, "\nError:", err)
+		return
+	}
+
+	//TODO move logic below somewhere close to Jira output implementation
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply msgservice/msghandling.go:45]: it's kind of technical debt

File: router/router.go (hunk 4 above)
Comment (AndreyLevchenko) [reply router/router.go:161]: done

File: router/router.go (hunk 3 above)
Comment (AndreyLevchenko) [reply router/router.go:91]: fixed

File: router/router.go (hunk 5 above)
Comment (AndreyLevchenko) [reply router/router.go:291]: - [ ] move variable definition before line `switch` statement

Comment (AndreyLevchenko) [reply router/router.go:291]: done

File: ui/backend/dbservice/getplgnstats.go (hunk 7 above)
Comment (AndreyLevchenko) [reply ui/backend/dbservice/getplgnstats.go:10]: yes.
In worth case stats won't be shown.

File: routes/routes.go (hunk 6 above)
Comment (AndreyLevchenko) [reply routes/routes.go:20]: no it won't. Route is not supposed to use after webhook termination

File: msgservice/logs.go [hunk 12]
```diff
@@ -3,12 +3,11 @@ package msgservice
 import "log"
 
 func prnInputLogs(msg string, v ...interface{}) {
+	maxLen := 20
```
Comment (afdesk) [inline msgservice/logs.go:6]: why do use a magic number - 20?

**Review by afdesk — CHANGES_REQUESTED** [review]

File: msgservice/logs.go (hunk 12 above)
Comment (AndreyLevchenko) [reply msgservice/logs.go:6]: the goal of entire function is to reduce amount of logging. If one of parameters passed is byte array then it's size is limited to 20 bytes, other parameters are go as is. I believe similar logic existed before:
https://github.com/aquasecurity/postee/commit/770ba7ef2ecf3d9ad9cd17f9619365fdbc75e1b7
(see changes in `logs.go`)

File: router/router.go [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+	mutexScan   sync.Mutex
+	quit        chan struct{}
+	queue       chan []byte
+	ticker      *time.Ticker
+	stopTicker  chan struct{}
+	cfgfile     string
+	aquaServer  string
+	outputs     map[string]outputs.Output
+	inputRoutes map[string]*routes.InputRoute
+	templates   map[string]data.Inpteval
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply router/router.go:44]: `outputs` and `templates` are interfaces

File: msgservice/scheduler.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+var getTicker = func(seconds int) *time.Ticker {
+	return time.NewTicker(time.Duration(seconds) * time.Second)
+}
+var RunScheduler = func(
+	route *routes.InputRoute,
+	fnSend func(plg outputs.Output, cnt map[string]string),
+	fnAggregate func(outputName string, currentContent map[string]string, counts int, ignoreLength bool) []map[string]string,
+	inpteval data.Inpteval,
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply msgservice/scheduler.go:19]: it's interface

File: dbservice/dbparam.go [hunk 15]
```diff
@@ -6,9 +6,11 @@ import (
 )
 
 var (
-	dbBucketName        = "WebhookBucket"
-	dbBucketAggregator  = "WebhookAggregator"
-	dbBucketExpiryDates = "WebookExpiryDates"
+	dbBucketName         = "WebhookBucket"
+	dbBucketAggregator   = "WebhookAggregator"
+	dbBucketExpiryDates  = "WebookExpiryDates"
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply dbservice/dbparam.go:11]: hm..
it's easy fix. But error was introduced before `v2`. It means there are customers who have bolt db with `WebookExpiryDates` bucket. If we change bucket name their buckets will be lost. If it's ok I can change bucket name

File: router/router.go (hunk 1 above)
Comment (AndreyLevchenko) [reply router/router.go:48]: fixed

File: router/router.go (hunk 2 above)
Comment (AndreyLevchenko) [reply router/router.go:53]: fixed

File: dbservice/plgnstats.go [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+			i = 0
+		} else {
+			i, err = strconv.Atoi(string(v[:]))
+		}
+
+		i++
+		nwv := strconv.Itoa(i)
+
+		err = bucket.Put([]byte(name), []byte(nwv))
+		return err
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply dbservice/plgnstats.go:38]: added fix to msgservice.go

File: dbservice/dbparam.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (afdesk) [reply dbservice/dbparam.go:11]: yes, it's my fault. let's not change it yet.

**Pull request merged**